### ISSUE TITLE
Resource Pack Loader Uses Unsorted List

### ISSUE THREAD

**Issue by Syrikal** [MinecraftForge/MinecraftForge#7644]
**Minecraft Version:** 1.16.4

**Forge Version:** 35.1.28

**Description of issue:**
I'm having issues with mod data not loading in the correct order. I asked about it [on the forums](https://forums.minecraftforge.net/topic/97313-116-how-to-ensure-load-order/) and was told that 'ResourcePackLoader.loadResourcePacks seems to use the unsorted mod list' and that I should file a bug report here.

**Comment by sciwhiz12** [comment]
I'm trying to debug this issue, and it seems like `ResourcePackLoader#loadResourcePacks` uses the sorted mod list.

#### _TL;DR: the mod sorter passes down sorted lists of both `ModFileInfo`s and `ModInfo`s, which is used by `ResourcePackLoader#loadResourcePacks`_
`ResourcePackLoader#loadResourcePacks` calls `ModList#getModFiles`, which is populated by `ModList#of`, which is called by `ModLoader#gatherAndInitializeMods`, which calls `LoadingModList#getModFiles`, which is populated by `LoadingModList#of`, which is called by `ModSorter#of`, which (if there are no errors) passes in the `ModSorter#modFiles` field from a `ModSorter` which has been sorted using `ModSorter#sort`, which sets the `modFiles` field from the `sorted` local variable (which is also used to set the `ModSorter#sortedList` field, which is passed down and down by the same chain).

This leads me to believe that your issue is due to some other condition in your situation. Without further information (even from the forums post), we cannot reasonably debug your issue.

However, I have a theory: your modid probably is before `minecraft` and `tetra` when sorted lexicographically/alphabetically, which means that your resource data is loaded before Minecraft's and Tetra's. A possible solution would be to [declare dependencies][fcw-dep-configs] on both Minecraft and Tetra in your `mods.toml`, and setting the `order` property to `AFTER` on both so your addon mod is sorted after the two in the mod list, which will propagate down to `ResourcePackLoader` (through the previously described call chain).

[fcw-dep-configs]: https://forge.gemwire.uk/wiki/Proper_Mod_Structuring#Dependency_Configurations

**Comment by LexManos** [comment]
Yes, default ordering of mod resources is respective of the mod sorted order.
This has been tried and tested many times.
However it's a known case that there are dumb mods out there that modify the resource list for whatever their personal reason may be.
So if your order is not correct/not in mod sort order. Start removing mods till you find the correct one.

**Comment by Syrikal** [comment]
Huh. I could've sworn the dependency-ordering method didn't work when I tried it last time, but I tried it again and it works fine! Sorry for the trouble, and thank you for the help!
